File: Cauly3.1/MediationSample/CaulyObjectiveCMediationSample/CaulyCustomEvent/CaulyEvent.m
//
//  CaulyEvent.m
//  CaulyIOSMediationSample
//
//  Created by FSN on 2023/03/15.
//

#import "CaulyEvent.h"
#import "CaulyEventBanner.h"
#import "CaulyEventInterstitial.h"
#import "CaulyEventNative.h"

@implementation CaulyEvent {
    CaulyEventBanner *caulyBanner;
    
    CaulyEventInterstitial *caulyInterstitial;
    
    CaulyEventNative *caulyNative;
}

+ (GADVersionNumber)adSDKVersion {
    NSArray *versionComponents = [CAULY_SDK_VERSION componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count >= 3) {
      version.majorVersion = [versionComponents[0] integerValue];
      version.minorVersion = [versionComponents[1] integerValue];
      version.patchVersion = [versionComponents[2] integerValue];
    }
    return version;
}

+ (GADVersionNumber)adapterVersion {
    NSString *adapterVersion = @"1.0.0.0";
    NSArray *versionComponents = [adapterVersion componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count == 4) {
    version.majorVersion = [versionComponents[0] integerValue];
    version.minorVersion = [versionComponents[1] integerValue];
    version.patchVersion =
        [versionComponents[2] integerValue] * 100 + [versionComponents[3] integerValue];
    }
    return version;
}

+ (nullable Class<GADAdNetworkExtras>)networkExtrasClass {
    return nil;
}


+ (void)setUpWithConfiguration:(GADMediationServerConfiguration *)configuration completionHandler:(GADMediationAdapterSetUpCompletionBlock)completionHandler {
    completionHandler(nil);
}

#pragma mark - Cauly Banner Ad Request
- (void)loadBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler {
    caulyBanner = [[CaulyEventBanner alloc] init];
    [caulyBanner loadBannerForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

#pragma mark - Cauly Interstitial Ad Request
- (void)loadInterstitialForAdConfiguration:
            (GADMediationInterstitialAdConfiguration *)adConfiguration
                         completionHandler:
                             (GADMediationInterstitialLoadCompletionHandler)completionHandler {
    caulyInterstitial = [[CaulyEventInterstitial alloc] init];
    [caulyInterstitial loadInterstitialForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

#pragma mark - Cauly Native Ad request
- (void)loadNativeAdForAdConfiguration:(GADMediationNativeAdConfiguration *)adConfiguration
                     completionHandler:(GADMediationNativeLoadCompletionHandler)completionHandler {
  caulyNative = [[CaulyEventNative alloc] init];
  [caulyNative loadNativeAdForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

@end

File: Cauly3.1/MediationSample/CaulySwiftMediationSample/CaulyCustomEvent/CaulyEvent.swift
//
//  CaulyEvent.swift
//  CaulySwiftMediationSample
//
//  Created by FSN on 2023/07/24.
//

import Foundation
import GoogleMobileAds

class CaulyEvent: NSObject, GADMediationAdapter {
    
    fileprivate var bannerAd: CaulyEventBanner?
    
    fileprivate var interstitialAd: CaulyEventInterstitial?
    
    fileprivate var nativeAd: CaulyEventNative?
    
    required override init() {
        super.init()
    }
    
    // MARK: - Cauly Banner Ad Request
    func loadBanner(for adConfiguration: GADMediationBannerAdConfiguration, completionHandler: @escaping GADMediationBannerLoadCompletionHandler) {
        self.bannerAd = CaulyEventBanner()
        self.bannerAd?.loadBanner(for: adConfiguration, completionHandler: completionHandler)
    }
    
    // MARK: - Cauly Interstitial Ad Request
    func loadInterstitial(for adConfiguration: GADMediationInterstitialAdConfiguration, completionHandler: @escaping GADMediationInterstitialLoadCompletionHandler) {
        self.interstitialAd = CaulyEventInterstitial()
        self.interstitialAd?.loadInterstitial(for: adConfiguration, completionHandler: completionHandler)
    }
    
    // MARK: - Cauly Native Ad Request
    func loadNativeAd(for adConfiguration: GADMediationNativeAdConfiguration, completionHandler: @escaping GADMediationNativeLoadCompletionHandler) {
        self.nativeAd = CaulyEventNative()
        self.nativeAd?.loadNativeAd(for: adConfiguration, completionHandler: completionHandler)
    }

    static func setUpWith(_ configuration: GADMediationServerConfiguration, completionHandler: @escaping GADMediationAdapterSetUpCompletionBlock) {
        // This is where you will initialize the SDK that this custom event is built
        // for. Upon finishing the SDK initialization, call the completion handler
        // with success.
        completionHandler(nil)
    }

    static func adapterVersion() -> GADVersionNumber {
        let adapterVersion = "1.0.0.0"
        let versionComponents = adapterVersion.components(separatedBy: ".")
        var version = GADVersionNumber()
        if versionComponents.count == 4 {
            version.majorVersion = Int(versionComponents[0]) ?? 0
            version.minorVersion = Int(versionComponents[1]) ?? 0
            version.patchVersion = (Int(versionComponents[2]) ?? 0) * 100 + (Int(versionComponents[3]) ?? 0)
        }
        return version
    }

    static func adSDKVersion() -> GADVersionNumber {
        let versionComponents = CAULY_SDK_VERSION.components(separatedBy: ".")
        
        if versionComponents.count >= 3 {
            let majorVersion = Int(versionComponents[0]) ?? 0
            let minorVersion = Int(versionComponents[1]) ?? 0
            let patchVersion = Int(versionComponents[2]) ?? 0
            
            return GADVersionNumber(majorVersion: majorVersion, minorVersion: minorVersion, patchVersion: patchVersion)
        }
        
        return GADVersionNumber()
    }

    static func networkExtrasClass() -> GADAdNetworkExtras.Type? {
        return nil
    }
}
